atomicfu-transformer/src/main/kotlin/kotlinx/atomicfu/transformer/AtomicFUTransformer.kt

```diff
@@ -785,33 +775,71 @@ class AtomicFUTransformer(
 
         private fun fixupLoadedAtomicVar(f: FieldInfo, ld: FieldInsnNode): AbstractInsnNode? {
             val j = FlowAnalyzer(ld.next).execute()
-            when (j) {
+            return fixupOperationOnAtomicVar(j, f, ld, null)
+        }
+
+        private fun fixupLoadedArrayElement(f: FieldInfo, ld: FieldInsnNode, getter: MethodInsnNode): AbstractInsnNode? {
+            // contains array field load (in vh case: + swap and pure type array load) and array element index
+            val arrayElementInfo = mutableListOf<AbstractInsnNode>()
+            if (vh) {
+                arrayElementInfo.add(ld.previous.previous) // getstatic VarHandle field
+                arrayElementInfo.add(ld.previous) // swap
+            }
+            var i: AbstractInsnNode = ld
+            while (i != getter) {
+                arrayElementInfo.add(i)
+                i = i.next
+            }
+            // start of array element operation arguments
+            val args = getter.next
+            // remove array element getter
+            instructions.remove(getter)
+            val arrayElementOperation = FlowAnalyzer(args).execute()
+            return fixupOperationOnAtomicVar(arrayElementOperation, f, ld, arrayElementInfo)
+        }
+
+        private fun fixupOperationOnAtomicVar(operation: AbstractInsnNode, f: FieldInfo, ld: FieldInsnNode, arrayElementInfo: List<AbstractInsnNode>?): AbstractInsnNode? {
+            val onArrayElement = f.getPrimitiveType(vh).sort == ARRAY
```

Comment:
It seems redundant now, as `onArrayElement` shoud be equivalent to`arrayElementInfo != null`
At least, an assertion to this fact would be helpful here:
```
check(onArrayElement == (arrayElementInfo != null))
```
However, when I add this kind of `check` it fails on `testIndexArrayElementGetting`. This is strage. It means that some loads are not being recognized as "array loads", so they work like regular functions, but will not work with `inline` function. 

It looks like the problem with primitive arrays with inline functions. Please, add corresponding tests.

Reply:
`fixupLoadedArrayElement` is called only when `get` array element instruction is visited: it gets information about array field and element index (`arrayElementInfo`) that will be used only in case of loading a stored reference to this array element. 
So, `arrayElementInfo` is not null only if operation on the array element is being processed but if this fixup is processing operation on the array field itself (only the `get(I)` element instruction can follow it ) then `arrayElementInfo` is still null. Though,  in both cases `f.getPrimitiveType(vh).sort == ARRAY`. 

I moved this check for `arrayElementInfo != null` to the branch of storing operation as it is only used in that case.